File: nukkit_binding/src/main/java/com/blackjack200/xyron/nukkit/Convert.java
package com.blackjack200.xyron.nukkit;

import cn.nukkit.Player;
import cn.nukkit.block.Block;
import cn.nukkit.block.BlockID;
import cn.nukkit.block.BlockLiquid;
import cn.nukkit.event.entity.EntityDamageEvent;
import cn.nukkit.item.*;
import cn.nukkit.math.AxisAlignedBB;
import cn.nukkit.math.SimpleAxisAlignedBB;
import cn.nukkit.math.Vector3;
import cn.nukkit.potion.Effect;
import com.github.blackjack200.xyron.PlayerOuterClass;
import com.github.blackjack200.xyron.PlayerWrappers;
import com.github.blackjack200.xyron.PrimitiveTypes;
import com.google.common.collect.Lists;
import lombok.experimental.UtilityClass;
import lombok.val;
import lombok.var;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

@UtilityClass
public class Convert {

    public PlayerOuterClass.DeviceOS deviceOS(int os) {
        var c = PlayerOuterClass.DeviceOS.forNumber(os);
        if (c == null) {
            c = PlayerOuterClass.DeviceOS.Android;
        }
        return c;
    }

    public PlayerOuterClass.GameMode gameMode(int gameMode) {
        switch (gameMode) {
            case Player.CREATIVE:
                return PlayerOuterClass.GameMode.Creative;
            case Player.ADVENTURE:
                return PlayerOuterClass.GameMode.Adventure;
            case Player.SPECTATOR:
                return PlayerOuterClass.GameMode.Spectator;
            case Player.SURVIVAL:
                return PlayerOuterClass.GameMode.Survival;
            default:
                throw new RuntimeException("invalid gamemode: " + gameMode);
        }
    }

    public PlayerOuterClass.DamageCause damageCause(EntityDamageEvent.DamageCause cause) {
        if (cause == EntityDamageEvent.DamageCause.CONTACT) {
            return PlayerOuterClass.DamageCause.Contact;
        }
        if (cause == EntityDamageEvent.DamageCause.ENTITY_ATTACK) {
            return PlayerOuterClass.DamageCause.EntityAttack;
        }
        if (cause == EntityDamageEvent.DamageCause.PROJECTILE) {
            return PlayerOuterClass.DamageCause.Projectile;
        }
        if (cause == EntityDamageEvent.DamageCause.SUFFOCATION) {
            return PlayerOuterClass.DamageCause.Suffocation;
        }
        if (cause == EntityDamageEvent.DamageCause.FALL) {
            return PlayerOuterClass.DamageCause.Fall;
        }
        if (cause == EntityDamageEvent.DamageCause.FIRE) {
            return PlayerOuterClass.DamageCause.Fire;
        }
        if (cause == EntityDamageEvent.DamageCause.FIRE_TICK) {
            return PlayerOuterClass.DamageCause.FireTick;
        }
        if (cause == EntityDamageEvent.DamageCause.LAVA) {
            return PlayerOuterClass.DamageCause.Lava;
        }
        if (cause == EntityDamageEvent.DamageCause.DROWNING) {
            return PlayerOuterClass.DamageCause.Drowning;
        }
        if (cause == EntityDamageEvent.DamageCause.BLOCK_EXPLOSION) {
            return PlayerOuterClass.DamageCause.BlockExplosion;
        }
        if (cause == EntityDamageEvent.DamageCause.ENTITY_EXPLOSION) {
            return PlayerOuterClass.DamageCause.EntityExplosion;
        }
        if (cause == EntityDamageEvent.DamageCause.VOID) {
            return PlayerOuterClass.DamageCause.Void;
        }
        if (cause == EntityDamageEvent.DamageCause.SUICIDE) {
            return PlayerOuterClass.DamageCause.Suicide;
        }
        if (cause == EntityDamageEvent.DamageCause.MAGIC) {
            return PlayerOuterClass.DamageCause.Magic;
        }
        if (cause == EntityDamageEvent.DamageCause.CUSTOM) {
            return PlayerOuterClass.DamageCause.Custom;
        }
        if (cause == EntityDamageEvent.DamageCause.LIGHTNING) {
            return PlayerOuterClass.DamageCause.Custom;
        }
        if (cause == EntityDamageEvent.DamageCause.HUNGER) {
            return PlayerOuterClass.DamageCause.Starvation;
        }
        return PlayerOuterClass.DamageCause.Custom;
    }

    public PlayerOuterClass.InputMode inputMode(int mode) {
        var c = PlayerOuterClass.InputMode.forNumber(mode);
        if (c == null) {
            c = PlayerOuterClass.InputMode.Touch;
        }
        return c;
    }

    public PrimitiveTypes.Vec3f vec3f(Vector3 vec3) {
        return PrimitiveTypes.Vec3f.newBuilder()
                .setX((float) vec3.getX())
                .setY((float) vec3.getY())
                .setZ((float) vec3.getZ())
                .build();
    }

    public Vector3 vec3f(PrimitiveTypes.Vec3f v3f) {
        return new Vector3(v3f.getX(), v3f.getY(), v3f.getZ());
    }

    public PrimitiveTypes.Vec3i vec3i(Vector3 vec3) {
        return PrimitiveTypes.Vec3i.newBuilder()
                .setX((int) vec3.getX())
                .setY((int) vec3.getY())
                .setZ((int) vec3.getZ())
                .build();
    }

    public Vector3 vec3i(PrimitiveTypes.Vec3i vec3) {
        return new Vector3(vec3.getX(), vec3.getY(), vec3.getZ());
    }

    public PrimitiveTypes.AxisAlignedBoundingBox boundingBox(AxisAlignedBB aabb) {
        return PrimitiveTypes.AxisAlignedBoundingBox.newBuilder()
                .setMin(vec3f(new Vector3(aabb.getMinX(), aabb.getMinY(), aabb.getMinZ())))
                .setMax(vec3f(new Vector3(aabb.getMaxX(), aabb.getMaxY(), aabb.getMaxZ())))
                .build();
    }

    public AxisAlignedBB boundingBox(PrimitiveTypes.AxisAlignedBoundingBox aabb) {
        return new SimpleAxisAlignedBB(vec3f(aabb.getMin()), vec3f(aabb.getMax()));
    }

    public PrimitiveTypes.BlockData block(Block blk) {
        return PrimitiveTypes.BlockData.newBuilder()
                .setPosition(vec3i(blk))
                .setFeature(PrimitiveTypes.BlockFeature.newBuilder()
                        .addAllCollisionBoxes(Lists.newArrayList(boundingBox(blk.getCollisionBoundingBox())))
                        .setFriction((float) blk.getFrictionFactor())
                        .setIsSolid(blk.isSolid())
                        .setIsLiquid(blk instanceof BlockLiquid)
                        .setIsAir(blk.getId() == BlockID.AIR)
                        .setIsSlime(blk.getId() == BlockID.SLIME_BLOCK)
                        .setIsClimbable(blk.canBeClimbed())
                        .setIsIce(blk.getId() == BlockID.ICE)
                        .setIsCobweb(blk.getId() == BlockID.COBWEB)
                        //no sweet berry in Nukkit
                        .setIsSweetBerry(false)
                        .build()
                ).build();
    }

    public PrimitiveTypes.ItemData item(Item item) {
        return PrimitiveTypes.ItemData.newBuilder()
                //FIXME unfortunately Nukkit don't use this
                .setVanillaName(item.getName())
                .setCount(item.getCount())
                .setFeature(PrimitiveTypes.ItemFeature.newBuilder()
                        .setIsArmor(item.isArmor())
                        .setIsBlockPlanterItem(false)
                        .setIsDamageable(item instanceof ItemDurable)
                        .setIsFood(item instanceof ItemEdible)
                        .setIsThrowable(item instanceof ProjectileItem)
                        .setIsTool(item.isTool())
                        .setIsBow(item instanceof ItemBow)
                        .setIsCrossBow(item instanceof ItemCrossbow)
                        .setIsShield(item instanceof ItemShield)
                        .build()
                )
                .build();
    }

    public List<PrimitiveTypes.EffectFeature> effects(Map<Integer, Effect> effects) {
        val l = new ArrayList<PrimitiveTypes.EffectFeature>(16);
        internalEffect(l, effects, Effect.SPEED, (e) -> e.setIsSpeed(true));
        internalEffect(l, effects, Effect.HASTE, (e) -> e.setIsHaste(true));
        internalEffect(l, effects, Effect.SLOW_FALLING, (e) -> e.setIsSlowFalling(true));
        internalEffect(l, effects, Effect.LEVITATION, (e) -> e.setIsLevitation(true));
        internalEffect(l, effects, Effect.SLOWNESS, (e) -> e.setIsSlowness(true));
        internalEffect(l, effects, Effect.JUMP_BOOST, (e) -> e.setIsJumpBoost(true));
        return l;
    }

    private void internalEffect(List<PrimitiveTypes.EffectFeature> list, Map<Integer, Effect> effects, int effectId, Consumer<PrimitiveTypes.EffectFeature.Builder> c) {
        val eff = effects.getOrDefault(effectId, null);
        if (eff != null) {
            val ef = PrimitiveTypes.EffectFeature.newBuilder()
                    .setAmplifier(eff.getAmplifier());
            c.accept(ef);
            list.add(ef.build());
        }
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerActionData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setActionData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerMoveData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setMoveData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerPlaceBlockData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setPlaceBlockData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerBreakBlockData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setBreakBlockData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerEatFoodData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setEatFoodData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerAttackData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setAttackData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerEffectData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setEffectData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerGameModeData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setGameModeData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerMotionData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setMotionData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerInputModeData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setInputModeData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerHeldItemChangeData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setHeldItemChangeData(data).build();
    }

    public PlayerWrappers.WildcardReportData wildcard(PlayerWrappers.PlayerLifeData data) {
        return PlayerWrappers.WildcardReportData.newBuilder().setLifeData(data).build();
    }
}

File: pmmp_binding/src/prokits/xyron/loader/Convert.php
<?php

namespace prokits\xyron\loader;

use Closure;
use pocketmine\block\Block;
use pocketmine\block\BlockTypeIds;
use pocketmine\block\Liquid;
use pocketmine\entity\effect\Effect;
use pocketmine\entity\effect\EffectManager;
use pocketmine\entity\effect\VanillaEffects;
use pocketmine\event\entity\EntityDamageEvent;
use pocketmine\item\Armor;
use pocketmine\item\Bow;
use pocketmine\item\Durable;
use pocketmine\item\Food;
use pocketmine\item\Item;
use pocketmine\item\ItemTypeIds;
use pocketmine\item\ProjectileItem;
use pocketmine\item\Tool;
use pocketmine\math\AxisAlignedBB;
use pocketmine\math\Vector3;
use pocketmine\player\GameMode;
use prokits\xyron\AxisAlignedBoundingBox;
use prokits\xyron\BlockData;
use prokits\xyron\BlockFeature;
use prokits\xyron\DamageCause;
use prokits\xyron\EffectFeature;
use prokits\xyron\GameMode as XyronGameMode;
use prokits\xyron\ItemData;
use prokits\xyron\ItemFeature;
use prokits\xyron\PlayerActionData;
use prokits\xyron\PlayerAttackData;
use prokits\xyron\PlayerBreakBlockData;
use prokits\xyron\PlayerEatFoodData;
use prokits\xyron\PlayerEffectData;
use prokits\xyron\PlayerGameModeData;
use prokits\xyron\PlayerHeldItemChangeData;
use prokits\xyron\PlayerInputModeData;
use prokits\xyron\PlayerLifeData;
use prokits\xyron\PlayerMotionData;
use prokits\xyron\PlayerMoveData;
use prokits\xyron\PlayerPlaceBlockData;
use prokits\xyron\Vec3f;
use prokits\xyron\Vec3i;
use prokits\xyron\WildcardReportData;

final class Convert {
	private function __construct() { }

	public static function deviceOS(int $os) : int {
		return $os;
	}

	public static function gameMode(GameMode $gameMode) : int {
		return match ($gameMode) {
			GameMode::CREATIVE() => XyronGameMode::Creative,
			GameMode::ADVENTURE() => XyronGameMode::Adventure,
			GameMode::SURVIVAL() => XyronGameMode::Survival,
			GameMode::SPECTATOR() => XyronGameMode::Spectator,
		};
	}

	public static function damageCause(int $cause) : int {
		return match ($cause) {
			EntityDamageEvent::CAUSE_CONTACT => DamageCause::Contact,
			EntityDamageEvent::CAUSE_ENTITY_ATTACK => DamageCause::EntityAttack,
			EntityDamageEvent::CAUSE_PROJECTILE => DamageCause::Projectile,
			EntityDamageEvent::CAUSE_SUFFOCATION => DamageCause::Suffocation,
			EntityDamageEvent::CAUSE_FALL => DamageCause::Fall,
			EntityDamageEvent::CAUSE_FIRE => DamageCause::Fire,
			EntityDamageEvent::CAUSE_FIRE_TICK => DamageCause::FireTick,
			EntityDamageEvent::CAUSE_LAVA => DamageCause::Lava,
			EntityDamageEvent::CAUSE_DROWNING => DamageCause::Drowning,
			EntityDamageEvent::CAUSE_BLOCK_EXPLOSION => DamageCause::BlockExplosion,
			EntityDamageEvent::CAUSE_ENTITY_EXPLOSION => DamageCause::EntityExplosion,
			EntityDamageEvent::CAUSE_VOID => DamageCause::Void,
			EntityDamageEvent::CAUSE_SUICIDE => DamageCause::Suicide,
			EntityDamageEvent::CAUSE_MAGIC => DamageCause::Magic,
			EntityDamageEvent::CAUSE_CUSTOM => DamageCause::Custom,
			EntityDamageEvent::CAUSE_STARVATION => DamageCause::Starvation
		};
	}

	public static function inputMode(int $mode) : int {
		return $mode;
	}

	public static function xyronVec3f(Vector3 $vec3) : Vec3f {
		return (new Vec3f())
			->setX((float) $vec3->getX())
			->setY((float) $vec3->getY())
			->setZ((float) $vec3->getZ());
	}

	public static function vec3f(Vec3f $v3f) : Vector3 {
		return new Vector3($v3f->getX(), $v3f->getY(), $v3f->getZ());
	}


	public static function xyronVec3i(Vector3 $vec3) : Vec3i {
		return (new Vec3i())
			->setX((int) $vec3->getX())
			->setY((int) $vec3->getY())
			->setZ((int) $vec3->getZ());
	}

	public static function vec3i(Vec3f $v3i) : Vector3 {
		return new Vector3($v3i->getX(), $v3i->getY(), $v3i->getZ());
	}

	public static function xyronBoundingBox(AxisAlignedBB $aabb) : AxisAlignedBoundingBox {
		return (new AxisAlignedBoundingBox())
			->setMin(self::xyronVec3f(new Vector3($aabb->minX, $aabb->minY, $aabb->minZ)))
			->setMax(self::xyronVec3f(new Vector3($aabb->maxX, $aabb->maxY, $aabb->maxZ)));
	}

	public static function boundingBox(AxisAlignedBoundingBox $aabb) : AxisAlignedBB {
		$min = $aabb->getMin();
		$max = $aabb->getMax();
		if ($min === null || $max === null) {
			throw new \RuntimeException();
		}
		return new AxisAlignedBB(
			$min->getX(), $min->getY(), $min->getZ(),
			$max->getX(), $max->getY(), $max->getZ(),
		);
	}

	public static function xyronBlockGetPosition(Block $blk) : BlockData {
		return self::xyronBlock($blk, $blk->getPosition());
	}

	public static function xyronBlock(Block $blk, Vector3 $position) : BlockData {
		$boxes = [];
		foreach ($blk->getCollisionBoxes() as $box) {
			$boxes[] = self::xyronBoundingBox($box);
		}
		return (new BlockData())
			->setPosition(self::xyronVec3i($position))
			->setFeature((new BlockFeature())
				->setCollisionBoxes($boxes)
				->setFriction((float) $blk->getFrictionFactor())
				->setIsSolid($blk->isSolid())
				->setIsLiquid($blk instanceof Liquid)
				->setIsAir($blk->getTypeId() === BlockTypeIds::AIR)
				->setIsSlime($blk->getTypeId() === BlockTypeIds::SLIME)
				->setIsClimbable($blk->canClimb())
				->setIsIce($blk->getTypeId() === BlockTypeIds::ICE)
				->setIsCobweb($blk->getTypeId() === BlockTypeIds::COBWEB)
				->setIsSweetBerry($blk->getTypeId() === BlockTypeIds::SWEET_BERRY_BUSH)
			);
	}

	public static function xyronItem(Item $item) : ItemData {
		return (new ItemData())
			->setVanillaName($item->getName())
			->setCount($item->getCount())
			->setFeature((new ItemFeature())
				->setIsArmor($item instanceof Armor)
				//TODO figure out what is this
				->setIsBlockPlanterItem(false)
				->setIsDamageable($item instanceof Durable)
				->setIsFood($item instanceof Food)
				->setIsThrowable($item instanceof ProjectileItem)
				->setIsTool($item instanceof Tool)
				->setIsBow($item instanceof Bow || $item->getTypeId() === ItemTypeIds::BOW)
				->setIsCrossBow($item->getTypeId() === ItemTypeIds::CROSSBOW)
				->setIsShield($item->getTypeId() === ItemTypeIds::SHIELD)
			);
	}

	/**
	 * @return EffectFeature[]
	 */
	public static function effects(EffectManager $effects) : array {
		$l = [];
		self::internalEffect($l, $effects, VanillaEffects::SPEED(), static fn(EffectFeature $e) => $e->setIsSpeed(true));
		self::internalEffect($l, $effects, VanillaEffects::HASTE(), static fn(EffectFeature $e) => $e->setIsHaste(true));
		//self::internalEffect($l, $effects, VanillaEffects::SLOW_FALLING(), static fn(EffectFeature $e) => $e->setIsSlowFalling(true));
		self::internalEffect($l, $effects, VanillaEffects::LEVITATION(), static fn(EffectFeature $e) => $e->setIsLevitation(true));
		self::internalEffect($l, $effects, VanillaEffects::SLOWNESS(), static fn(EffectFeature $e) => $e->setIsSlowness(true));
		self::internalEffect($l, $effects, VanillaEffects::JUMP_BOOST(), static fn(EffectFeature $e) => $e->setIsJumpBoost(true));
		return $l;
	}

	private static function internalEffect(array &$ls, EffectManager $effects, Effect $effect, Closure $c) : void {
		$eff = $effects->get($effect);
		if ($eff !== null) {
			$ef = (new EffectFeature())
				->setAmplifier($eff->getAmplifier());
			$c($ef);
			$ls[] = $ef;
		}
	}

	/**
	 * @param PlayerActionData|PlayerMoveData|PlayerPlaceBlockData|PlayerBreakBlockData|PlayerEatFoodData|PlayerAttackData|PlayerEffectData|PlayerGameModeData|PlayerMotionData|PlayerInputModeData|PlayerHeldItemChangeData|PlayerLifeData $data
	 */
	public static function wildcard($data) : WildcardReportData {
		if ($data instanceof PlayerActionData) {
			return (new WildcardReportData())->setActionData($data);
		}

		if ($data instanceof PlayerMoveData) {
			return (new WildcardReportData())->setMoveData($data);
		}

		if ($data instanceof PlayerPlaceBlockData) {
			return (new WildcardReportData())->setPlaceBlockData($data);
		}

		if ($data instanceof PlayerBreakBlockData) {
			return (new WildcardReportData())->setBreakBlockData($data);
		}

		if ($data instanceof PlayerEatFoodData) {
			return (new WildcardReportData())->setEatFoodData($data);
		}

		if ($data instanceof PlayerAttackData) {
			return (new WildcardReportData())->setAttackData($data);
		}

		if ($data instanceof PlayerEffectData) {
			return (new WildcardReportData())->setEffectData($data);
		}

		if ($data instanceof PlayerGameModeData) {
			return (new WildcardReportData())->setGameModeData($data);
		}

		if ($data instanceof PlayerMotionData) {
			return (new WildcardReportData())->setMotionData($data);
		}

		if ($data instanceof PlayerInputModeData) {
			return (new WildcardReportData())->setInputModeData($data);
		}

		if ($data instanceof PlayerHeldItemChangeData) {
			return (new WildcardReportData())->setHeldItemChangeData($data);
		}

		if ($data instanceof PlayerLifeData) {
			return (new WildcardReportData())->setLifeData($data);
		}
		throw new \RuntimeException();
	}
}